### PR TITLE
added mission nod4a

### PR DESCRIPTION
**Pull request by 42foobar42**
After a break, here we go again.

I think there are a few points that can be discussed on this  mission.

So happy testing and I am waiting for your comments.

### CODE REVIEW

File: mods/cnc/maps/nod04a/nod04a.lua [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+	end)
+end
+
+IdleHunt = function(unit) 
+	if not unit.IsDead then 
+		Trigger.OnIdle(unit, unit.Hunt) 
+	end 
+end
+
+getActors = function(owner, units)
```
Comment (Mailaender) [inline mods/cnc/maps/nod04a/nod04a.lua]: This should make use of the new Lua API function from https://github.com/OpenRA/OpenRA/pull/7463.

File: mods/cnc/maps/nod04a/map.yaml [hunk 2]
```diff
@@ -0,0 +1,741 @@
...
+		OwnsWorld: True
+		NonCombatant: True
+		Race: gdi
+		Allies: Nod
+		Enemies: GDI
+	PlayerReference@GDI:
+		Name: GDI
+		Race: gdi
+		ColorRamp: 31,222,183
+		Allies: GDI
```
Comment (abcdefg30) [inline mods/cnc/maps/nod04a/map.yaml]: You don't need this here.

**Comment by Mailaender** [comment]
The building husks don't hide the owner row:

image

**Comment by Mailaender** [comment]
Winning and Loosing scenario seem to work properly.

**Comment by 42foobar42** [comment]
I keep in mind to change the the discover trigger as soon as it is merged.

Winning scenario is killing all civilains in the north west village. It does not matter if you killed all GDI infantry.

Loosing scenrario is if you loose all your own infantry or all civilians in the south east village are killed.

**Comment by Mailaender** [comment]
https://github.com/OpenRA/OpenRA/pull/7492 has been merged.

**Comment by penev92** [comment]
I noted this on IRC at the time, but I was really close to asking about progress here because I had forgotten myself: the `Has dependencies` label was for #7463, not #7492. (or maybe both; point is #7463 isn't merged, so the label should come back).
Anyway, have you made any progress on this, @42foobar42 ?

File: mods/cnc/maps/nod04a/nod04a.lua [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+InsertNodUnits = function()
+	Reinforcements.Reinforce(Nod, NodUnits, { UnitsEntry.Location, UnitsRally.Location }, 15)
+end
+
+CreateCivilians = function(discovered, discoverer)
+	Utils.Do(NodCivilians, function(civilian)
+		local Actor = Reinforcements.Reinforce(Nod, civilian[1], civilian[2], 15, nil)
+		NodCiviliansActors[#NodCiviliansActors+1] = Actor[1]
+	end)
+	NodObjective2 = Nod.AddPrimaryObjective("Protect the civilians that supports Nod.")
```
Comment (obrakmann) [inline mods/cnc/maps/nod04a/nod04a.lua]: Typo: "[...] that *support* Nod."

File: mods/cnc/maps/nod04a/nod04a.lua [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+ShowCivilianCellTriggerActivator = { CPos.New(43,29), CPos.New(43,30), CPos.New(43,31), CPos.New(43,32), CPos.New(43,33), CPos.New(43,34), CPos.New(43,35), CPos.New(43,36), CPos.New(43,37), CPos.New(43,38), CPos.New(43,39) }
+
+Atk3TriggerCounter = 2
+
+Atk1TriggerFunctionTime = DateTime.Seconds(20)
+XxxxTriggerFunctionTime = DateTime.Seconds(50)
+YyyyTriggerFunctionTime = DateTime.Minutes(1) + DateTime.Seconds(40)
+ZzzzTriggerFunctionTime = DateTime.Minutes(2) + DateTime.Seconds(30)
+
+NodCivilians = { { { 'c9' }, { spawnciv0.Location } }, { { 'c8' }, { spawnciv1.Location } }, { { 'c7' }, { spawnciv2.Location } }, { { 'c4' }, { spawnciv3.Location } }, { { 'c3' }, { spawnciv4.Location } }, { { 'c1' }, { spawnciv5.Location } }, { { 'c4' }, { spawnciv6.Location } }, { { 'c5' }, { spawnciv7.Location } }, { { 'c2' }, { spawnciv8.Location } } }
```
Comment (obrakmann) [inline mods/cnc/maps/nod04a/nod04a.lua]: These just popping up at those locations looks a bit weird. Is it possible to find another solution? Maybe have them placed in the map from the start and only change the owner when the trigger activates? Or let them come out of their houses or something.

File: mods/cnc/maps/nod04a/nod04a.lua [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+
+Atk4Movement = function(unit)
+	Utils.Do(Atk4Waypoints, function(waypoint)
+		unit.AttackMove(waypoint.Location)
+	end)
+	IdleHunt(unit)
+end
+
+InsertNodUnits = function()
+	Reinforcements.Reinforce(Nod, NodUnits, { UnitsEntry.Location, UnitsRally.Location }, 15)
```
Comment (obrakmann) [inline mods/cnc/maps/nod04a/nod04a.lua]: Could you split the entry and rallypoint locations for those a bit? It takes ages for all those units to enter, and they all huddle up around the same place, which doesn't look too good.

**Comment by obrakmann** [comment]
Gameplay-wise this is fine, minus the mentioned nits. Knowing our audience, it's probably a bit hard. I don't have any good idea how to implement difficulty levels for it though, given how the mission works.

I'll look at the code when this has been rebased against #7463.

**Comment by obrakmann** [comment]
Oh, I almost forgot to mention: there's this graphical glitch, although that's most likely our fault:

screenshot from 2015-04-18 18 22 33

**Comment by 42foobar42** [comment]
Hi there,
I fixed the mention points.
But the trigger OnPlayerDiscovered did not work. There is no error or something. I could not figure out what the problem is. I try to do it the same way like gdi mission 5. 	Obviously I overlook something.

File: mods/cnc/maps/nod04a/map.yaml [hunk 6]
```diff
@@ -0,0 +1,709 @@
...
+	UnitsEntryRocket: waypoint
+		Location: 56,12
+		Owner: Neutral
+	UnitsRallyRocket: waypoint
+		Location: 54,12
+		Owner: Neutral
+	NodCiv1: c8
+		Location: 46,36
+		Owner: NodSupporter
+		Health: 1
```
Comment (abcdefg30) [inline mods/cnc/maps/nod04a/map.yaml]: Could you please remove those `Health: 1` values here and below?

**Comment by abcdefg30** [comment]
Did you add `EnemyWatcher:` and `AnnounceOnSeen:`?

**Comment by Mailaender** [comment]
screenshot from 2015-04-18 18 22 33

> Oh, I almost forgot to mention: there's this graphical glitch, although that's most likely our fault:

This can be fixed by re-ordering the actors in the map.yaml listing.

**Comment by 42foobar42** [comment]
Thanks @abcdefg30 ! EnemyWatcher: and AnnounceOnSeen fixed the problem.
The other points are also fixed.

File: mods/cnc/maps/nod04a/map.yaml [hunk 7]
```diff
@@ -0,0 +1,705 @@
...
+	sc6 37,24 0:
+	sc6 36,18 0:
+
+Rules:
+	Player:
+		-ConquestVictoryConditions:
+		MissionObjectives:
+			EarlyGameOver: true
+		EnemyWatcher:
+			NotificationInterval: 25
```
Comment (abcdefg30) [inline mods/cnc/maps/nod04a/map.yaml]: Note: This is ok for now, but should be removed once #7984 is merged (or the other way round).

**Comment by penev92** [comment]
To summarize: this has dependencies on #7463 and #7984. Sorry @42foobar42

**Comment by abcdefg30** [comment]
#7984 is merged now.

**Comment by 42foobar42** [comment]
I removed the NotificationInterval, but I think it does not work as expected.

When my units arrive the village the trigger does not work.

File: mods/cnc/maps/nod04a/map.yaml [hunk 8]
```diff
@@ -0,0 +1,704 @@
...
+Smudges:
+	sc6 37,24 0:
+	sc6 36,18 0:
+
+Rules:
+	Player:
+		-ConquestVictoryConditions:
+		MissionObjectives:
+			EarlyGameOver: true
+		EnemyWatcher:
```
Comment (abcdefg30) [inline mods/cnc/maps/nod04a/map.yaml]: > When my units arrive the village the trigger does not work.

Indeed it does, but it takes a while. To fix this,
you probably want add `ScanInterval: 10` below this line (or maybe even a minor number).

And: Unintentional tab after the `:`.

File: mods/cnc/maps/nod04a/map.yaml [hunk 9]
```diff
@@ -0,0 +1,704 @@
...
+		Location: 50,37
+		Owner: NodSupporter
+		Facing: 96
+		SubCell: 2
+	NodCiv9: c9
+		Location: 47,34
+		Owner: NodSupporter
+		Facing: 192
+		SubCell: 2	
+
```
Comment (abcdefg30) [inline mods/cnc/maps/nod04a/map.yaml]: Unwanted tab here.

File: mods/cnc/maps/nod04a/map.yaml [hunk 10]
```diff
@@ -0,0 +1,704 @@
...
+	NodCiv8: c5
+		Location: 50,37
+		Owner: NodSupporter
+		Facing: 96
+		SubCell: 2
+	NodCiv9: c9
+		Location: 47,34
+		Owner: NodSupporter
+		Facing: 192
+		SubCell: 2
```
Comment (abcdefg30) [inline mods/cnc/maps/nod04a/map.yaml]: Same here.

**Comment by abcdefg30** [comment]
nod04a
This unit will stay at its spawn. Would be nice if you could do something agianst that.

File: mods/cnc/maps/nod04a/map.yaml (hunk 8 above)
Comment (penev92) [inline mods/cnc/maps/nod04a/map.yaml]: What is the problem here? The default rescan interval should be ~1 second, which is a compromise between discovery speed and performance, but generally should be good enough for this.

**Comment by penev92** [comment]
#7463 has been merged and I believe there are no more dependencies now.

File: mods/cnc/maps/nod04a/map.yaml (hunk 8 above)
Comment (abcdefg30) [inline mods/cnc/maps/nod04a/map.yaml]: Hm, yeah it is really not the rescan interval. I guess I'll send a patch myself, and clean some other things up as well.

**Comment by abcdefg30** [comment]
I think, I fixed everything in https://github.com/42foobar42/OpenRA/pull/4.

**Comment by 42foobar42** [comment]
Thanks @abcdefg30 . I merged your request.

Looks good to me now.

File: mods/cnc/maps/nod04a/nod04a.lua [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+		actor.Owner = Nod
+	end)
+
+	NodObjective2 = Nod.AddPrimaryObjective("Protect the civilians that support Nod.")
+	Trigger.OnAllKilled(NodCiviliansActors, function()
+		Nod.MarkFailedObjective(NodObjective2)
+	end)
+
+	Utils.Do(GcivActors, function(actor)
+		actor.AttackMove(waypoint7.Location)
```
Comment (abcdefg30) [inline mods/cnc/maps/nod04a/nod04a.lua]: Damn it. I'm that silly.
We need an `IsDead` check here.
```lua
Utils.Do(GcivActors, function(actor)
	if not actor.IsDead then
		actor.AttackMove(waypoint7.Location)
		actor.AttackMove(waypoint8.Location)
		IdleHunt(actor)
	end
end)

**Comment by abcdefg30** [comment]
Otherwise :+1: / :white_check_mark:

**Comment by 42foobar42** [comment]
no problem. I fixed it.

**Comment by Mailaender** [comment]
Works fine (I won, jay). :white_check_mark:

**Comment by Mailaender** [comment]
Thanks!

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/7e447a50da156681ac3b4910a5fcccd546978fc4...ba2a91cc699b6dc8ee644a036937ec56ef928bb5)

**Pull request merged**